Tracing: Support remote, rate-limited, and probabilistic sampling in tracing.opentelemetry config section (#73587)
grafana/grafana: docs/sources/setup-grafana/configure-grafana/_index.md

Can be set with the environment variable `JAEGER_SAMPLER_TYPE`.

> _Setting `sampler_type` in the `tracing.opentelemetry` section will override this setting._

Reviewer: ```suggestion
_To override this setting, enter `sampler_type` in the `tracing.opentelemetry` section._
```

Reviewer: @hairyhenderson , wanted to check with you about the above suggestion before I replicate that wording convention to the docs below.

Can be set with the environment variable `OTEL_RESOURCE_ATTRIBUTES` (use `=` instead of `:` with the environment variable).

### sampler_type

Default value is `const`.

> Specifies the type of sampler: `const`, `probabilistic`, `ratelimiting`, or `remote`.

Reviewer: ```suggestion
This option specifies the type of sampler: `const`, `probabilistic`, `ratelimiting`, or `remote`.
```

Author: hmm... is it "option" or "setting"? I see both used in a few places in this doc, but most entries are just implicit (that's why I used the original form)

Reviewer: Good point, it's probably "setting" but we can leave this as is so as not to introduce an inconsistency in the docs. I'm fine with "Specifies the type...".

Author: ok - TBH there's already inconsistency - maybe that's worth cleaning up _later_ 😉

I'll leave this one alone for now.

Reviewer: 👍

### sampler_param

Default value is `1`.

> Sampler configuration parameter. Depending on the value of `sampler_type`, it can be `0`, `1`, or a decimal value in between.

Reviewer: ```suggestion
Depending on the value of `sampler_type`, the sampler configuration parameter can be `0`, `1`, or any decimal value between `0` and `1`.
```

> - For `const` sampler, `0` or `1` to always `false`/`true` respectively

Reviewer: ```suggestion
- For `const` sampler, `0` or `1` means `false`/`true` respectively
```

Reviewer: Does that suggestion  ^ make sense?

Author: yep - the "always" wording was perhaps trying to be too succinct. What was intended is that the sampling decision would always be `false` or `true` depending on the value.

There's probably a better way to word it. How about:

```suggestion
- For `const` sampler, use `0` to never sample or `1` to always sample
```

Reviewer: 👍 , that works for me, thank you.

> - For `probabilistic` sampler, a probability between `0.0` and `1.0`

Reviewer: ```suggestion
- For the `probabilistic` sampler, you can use a decimal value between `0.0` and `1.0`
```

> - For `rateLimiting` sampler, the number of spans per second

Reviewer: ```suggestion
- For the `rateLimiting` sampler, enter the number of spans per second
```

> - For `remote` sampler, param is the same as for `probabilistic`

Reviewer: ```suggestion
- For the `remote` sampler, enter the same parameter value as you entered for `probabilistic`.
```

Author: This doesn't make sense - the new wording implies the value should match what was entered for `probabilistic`, except it's an either-or choice - you can't choose both the `remote` _and_ `probabilistic` sampler.

How about:

```suggestion
- For the `remote` sampler, use a decimal value between `0.0` and `1.0` to specify
```

Author: See https://github.com/grafana/grafana/pull/73587/files#r1320166327

Reviewer: Ah, that works for me. Thank you.

>   and indicates the initial sampling rate before the actual one

Reviewer: ```suggestion
 the initial sampling rate before a sample is received from the sampling server.
```

Author: This no longer makes sense as samples are not received from the sampling server...

Also, a multi-line suggestion would be _much_ easier to reason through 😅 I'm going to start a new one with a counter-proposal.

Author: https://github.com/grafana/grafana/pull/73587/files#r1320166327

>   and indicates the initial sampling rate before the actual one

Reviewer: ```suggestion
  and indicate the initial sampling rate before the actual one
```

> - For `remote` sampler, param is the same as for `probabilistic`
>   and indicates the initial sampling rate before the actual one
>   is received from the sampling server

Author: @chri2547 how about this:

```suggestion
- For the `remote` sampler, use a decimal value between `0.0` and `1.0`
  to specify the initial sampling rate used before the first update
  is received from the sampling server
```

Reviewer: Perfect, thank you.

### sampling_server_url

When `sampler_type` is `remote`, this specifies the URL of the sampling server. This can be used by all tracing providers. 

> The sampling server should implement the Jaeger remote sampling API, which can be exposed by a jaeger-agent, jaeger-collector, opentelemetry-collector-contrib, or the Grafana Agent.

Reviewer: does the user do this implementation? I'm wondering if this should be something like "In the sampling server, implement the Jaeger remote sampling API...".

Author: No, the user just needs to choose a server that implements that API - the 4 listed all implement the Jaeger remote sampling API.

Reviewer: 👍 , all good on this one then.

Author: How about:

```suggestion
Use a sampling server that implement the Jaeger remote sampling API, such as jaeger-agent, jaeger-collector, opentelemetry-collector-contrib, or Grafana Agent.
```

(maybe `Grafana Agent` should link to https://grafana.com/oss/agent/?)

Author: @chri2547 I think I was busy writing this comment and didn't see your original reply - what do you think of the alternate phrasing?

Reviewer: I like that, tho would use "to" instead of "that" - so "Use a sampling server to implement..."

Author: oops - it should've been `that implements`

"Use a sampling server to implement ..." doesn't really scan well to me - the user's really just running some server and putting its URL in this setting. The important thing is the server needs to support the correct API. Actually, maybe "supports" is a clearer term. Or maybe I'm bikeshedding myself at this point 😂

Author: How about:

```suggestion
Use a sampling server that supports the Jaeger remote sampling API, such as jaeger-agent, jaeger-collector, opentelemetry-collector-contrib, or [Grafana Agent](https://grafana.com/oss/agent/).
```

Reviewer: 👍

<hr>

> ## [tracing.opentelemetry.jaeger]

Author: huh, TIL 😂
